Shell/Twee.cpp

```diff
@@ -0,0 +1,219 @@
+ *  - introduced definitions are considered derived from the conjecture
+ */
+
+#include "Kernel/Clause.hpp"
+#include "Kernel/Problem.hpp"
+#include "Kernel/TermIterators.hpp"
+#include "Kernel/TermTransformer.hpp"
+#include "Kernel/Renaming.hpp"
+
+#include "Twee.hpp"
+
+using Kernel::Clause;
+using Kernel::Literal;
+using Kernel::Problem;
+using Kernel::Unit;
+using Kernel::UnitList;
+
+/*
+ * The actual worker for the twee trick.
+ # - evaluates conjucture literals bottom up (as orchestrated by Twee::apply)
+ # - and eagerly introduces new definitions for its subterms
+ # - already encountered subterms reuse older definitions
+ */
+class Definizator : public BottomUpTermTransformer {
+  public: // so that Twee::apply can directly access
+    // all the new definitions (as clauses) introduced along the way
+    UnitList* newUnits;
+
+    // accumlating the defining units as premises for the current "rewrite"
+    // (reset responsibility is with Twee::apply)
+    UnitList* premises;
+
+    // for each relevant term, cache the introduced symbol and the corresponding definition
+    DHMap<Term*,std::pair<unsigned,Clause*>> _cache;
+
+    Definizator(bool groundOnly) : newUnits(UnitList::empty()), _groundOnly(groundOnly) {}
+  private:
+    bool _groundOnly;
+
+    unsigned _typeArity;
+    TermStack _allVars; // including typeVars, which will come first, then termVars
+    TermStack _termVarSorts;
+
+    // a helper function to collect terms variables and their sorts
+    // all stored in the above private fields to be looked up by transformSubterm
+    void scanVars(Term* t) {
+      CALL("Definizator::scanVars");
+
+      static DHSet<unsigned> varSeen;
+      varSeen.reset();
+      _typeArity = 0;
+      _allVars.reset();
+      static TermStack termVars;
+      termVars.reset();
+      _termVarSorts.reset();
+
+      // fake scanVars cheaply
+      if (t->ground()) return;
+
+      VariableWithSortIterator it(t);
+      while (it.hasNext()) {
+        auto varAndSort = it.next();
+        unsigned v = varAndSort.first.var();
+        TermList s = varAndSort.second;
+        if (varSeen.insert(v)) {
+          if(s == AtomicSort::superSort()) {
+            _allVars.push(TermList(v,false));
+          } else {
+            termVars.push(TermList(v,false));
+            _termVarSorts.push(varAndSort.second);
+          }
+        }
+      }
+      _typeArity = _allVars.size(); // allVars only collected typeVars until now
+      for(unsigned i = 0; i < termVars.size(); i++){
+        _allVars.push(termVars[i]);
+      }
+
+      ASS_EQ(_typeArity+_termVarSorts.size(), _allVars.size())
+    }
+
+  protected:
+    TermList transformSubterm(TermList trm) override {
+      CALL("Definizator::transformSubterm");
+
+      // cout << "tf: " << trm.toString() << endl;
+      if (trm.isVar()) return trm;
+      Term* t = trm.term();
+      if (t->isSort() || t->arity() == 0 || (!t->ground() && _groundOnly)) return trm;
```

Comment:
One more thing: did we try also replacing constants? I'm fairly sure it doesn't make sense to replace sorts but in principle we could for HOL or something.

Reply:
I am fairly convinced that defining constants would only make it mildly slower with the default orderings... Defining sorts is not an option. Even the poly version of fol does not allow equalities between sorts to be stated!